### PR TITLE
engine, backtester: propagate runtime lifecycle and harden graceful shutdown

### PR DESCRIPTION
**Pull request by thrasher-**
## Summary

This PR hardens runtime/shutdown lifecycle handling across engine + backtester, and finalises the noctx/lifecycle workstream.

### Key changes

- Propagates a runtime-scoped cancellable context through startup/runtime paths.
- Replaces noctx listener usage with context-aware listener setup.
- Improves gRPC shutdown behavior:
  - graceful stop with timeout fallback
  - suppresses expected server-stopped shutdown noise
  - avoids concurrent shutdown call hangs via non-blocking shutdown signaling
- Fixes websocket routine manager start/stop race conditions around cancel/shutdown channels.
- Prevents subsystem re-enable during active shutdown lifecycle.
- Fixes order-manager shutdown cancellation flow so `CancelOrdersOnShutdown` executes correctly in real stop paths.
- Improves data-history shutdown semantics so context cancellation is treated as shutdown/abort (not persisted as misleading failed job outcomes).
- Adds/updates targeted regression tests for the above paths.

## Why

This closes the noctx lifecycle gap and addresses shutdown race/edge-case behavior observed during review:
- startup/shutdown context consistency
- concurrent shutdown safety
- websocket manager state/race correctness
- cleaner shutdown behavior under cancellation

## Validation

Ran locally:

- `go test ./engine -count=1`
- `go test ./backtester/engine -count=1`
- `go test -race ./engine -run TestWebsocketRoutineManagerConcurrentStartStop -count=1`
- `go test -race ./engine -run TestShutdownConcurrentRequestsDoNotBlock -count=1`
- `golangci-lint run ./engine/... ./backtester/engine/...`
- `codespell` on touched files
- Simulated `.github/workflows/misc.yml` checks (pass)

## Notes

- Patch is intentionally squashed for review clarity.
- Includes both production fixes and regression coverage for race/shutdown behavior.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -143,21 +144,37 @@ func main() {
 
 	engine.Bot.Settings.PrintLoadedSettings()
 
+	engine.Bot.EnsureRuntimeContext()
+	var shutdownRequested atomic.Bool
+	go waitForInterrupt(settings.Shutdown, engine.Bot, &shutdownRequested)
+
 	if err = engine.Bot.Start(); err != nil {
+		if shutdownRequested.Load() {
+			engine.Bot.Stop()
```
[the remark this answers is not in the document]
Comment (thrasher-) [reply main.go]: Jarvis: Great catch — addressed in commit https://github.com/thrasher-/gocryptotrader/commit/3e8a0c3bf23b916daada0dcecda074bfad05d33e. The startup-abort path in `main.go` now returns early when shutdown was requested during `Start()`, and no longer calls `Stop()` on a potentially partially initialised engine.

File: backtester/main.go [hunk 2]
```diff
@@ -218,19 +219,27 @@ func main() {
...
-			fmt.Printf("Could not start RPC server. Error: %v\n", err)
+		s, setupErr := backtest.SetupRPCServer(c, runManager)
+		if setupErr != nil {
+			fmt.Printf("Could not setup RPC server. Error: %v\n", setupErr)
+			os.Exit(1)
+		}
+		startErr := backtest.StartRPCServer(ctx, s)
+		if startErr != nil && ctx.Err() == nil {
+			fmt.Printf("Could not start RPC server. Error: %v\n", startErr)
 			os.Exit(1)
```
[the remark this answers is not in the document]
Comment (thrasher-) [reply backtester/main.go]: Jarvis: Agreed, and updated in commit https://github.com/thrasher-/gocryptotrader/commit/3e8a0c3bf23b916daada0dcecda074bfad05d33e. `backtester/main.go` no longer calls `os.Exit(1)` from inside the RPC goroutine; RPC setup/start errors are sent over an error channel and handled from the main goroutine so cancellation/cleanup occurs before process exit.

File: engine/datahistory_manager.go [hunk 3]
```diff
@@ -716,16 +722,19 @@ func (m *DataHistoryManager) processCandleData(job *DataHistoryJob, exch exchang
...
 		job.Asset,
 		job.Interval,
 		startRange,
 		endRange)
 	if err != nil {
+		if ctxErr := ctx.Err(); ctxErr != nil && errors.Is(err, ctxErr) {
+			return nil, err
+		}
 		r.Result += "could not get candles: " + err.Error() + ". "
 		r.Status = dataHistoryStatusFailed
```
[the remark this answers is not in the document]
Comment (thrasher-) [reply engine/datahistory_manager.go:736]: Jarvis: Addressed in commit https://github.com/thrasher-/gocryptotrader/commit/3e8a0c3bf23b916daada0dcecda074bfad05d33e. For context-cancellation cases, the path now returns the cancellation error (`return nil, err`) so shutdown/abort is explicit. For non-cancellation API errors, existing data-history behavior is preserved by recording failure details in the job result and returning `r, nil` intentionally.

File: engine/datahistory_manager.go [hunk 4]
```diff
@@ -769,15 +778,18 @@ func (m *DataHistoryManager) processTradeData(job *DataHistoryJob, exch exchange
...
 		startRange,
 		endRange)
 	if err != nil {
+		if ctxErr := ctx.Err(); ctxErr != nil && errors.Is(err, ctxErr) {
+			return nil, err
+		}
 		r.Result += "could not get trades: " + err.Error() + ". "
 		r.Status = dataHistoryStatusFailed
-		return r, nil //nolint:nilerr // error is returned in the job result
+		return r, nil
```
[the remark this answers is not in the document]
Comment (thrasher-) [reply engine/datahistory_manager.go:792]: Jarvis: Same treatment here in commit https://github.com/thrasher-/gocryptotrader/commit/3e8a0c3bf23b916daada0dcecda074bfad05d33e. Cancellation now returns a real context error; non-cancellation fetch errors continue to be encoded in the job result (`r`) by design to preserve existing data-history semantics.

File: backtester/engine/grpcserver.go [hunk 5]
```diff
@@ -86,22 +90,40 @@ func StartRPCServer(server *GRPCServer) error {
...
+			log.Errorln(log.GRPCSys, serveErr)
 			return
 		}
 	}()
 
 	log.Debugln(log.GRPCSys, "GRPC server started!")
 
 	if server.config.GRPC.GRPCProxyEnabled {
-		return server.StartRPCRESTProxy()
+		return server.startRPCRESTProxy(ctx)
```
Comment (shazbert) [inline backtester/engine/grpcserver.go]: If the proxy server fails for any reason, either initialise the proxy before starting Serve, or stop and close the gRPC server before returning the proxy error.

File: backtester/engine/grpcserver.go [hunk 6]
```diff
@@ -123,15 +145,25 @@ func (s *GRPCServer) StartRPCRESTProxy() error {
...
+		if shutdownErr := server.Shutdown(shutdownCtx); shutdownErr != nil && !errors.Is(shutdownErr, http.ErrServerClosed) {
+			log.Errorf(log.GRPCSys, "GRPC proxy server shutdown failed: %s\n", shutdownErr)
 		}
+	}()
 
-		if err = server.ListenAndServe(); err != nil {
-			log.Errorf(log.GRPCSys, "GRPC proxy failed to server: %s\n", err)
+	go func() {
+		if serveErr := server.ListenAndServe(); serveErr != nil && !errors.Is(serveErr, http.ErrServerClosed) {
+			log.Errorf(log.GRPCSys, "GRPC proxy failed to server: %s\n", serveErr)
```
Comment (shazbert) [inline backtester/engine/grpcserver.go]: Across these logs, the `\n` can be removed as it gets screened in the log package.

File: backtester/main.go [hunk 7]
```diff
@@ -218,51 +219,71 @@ func main() {
...
-	interrupt := <-signaler.WaitForInterrupt()
-	log.Infof(log.Global, "Captured %v, shutdown requested\n", interrupt)
-	if btCfg.StopAllTasksOnClose {
-		log.Infoln(log.Global, "Stopping all running tasks on close")
-		var stopped []*backtest.TaskSummary
-		stopped, err = runManager.StopAllTasks()
-		if err != nil {
-			log.Errorln(common.Backtester, err)
+		startErr := backtest.StartRPCServer(ctx, s)
+		if startErr != nil && ctx.Err() == nil {
```
Comment (shazbert) [inline backtester/main.go:233]: maybe if the context is erroring we should send that through the channel? as an else if case?

File: engine/engine.go [hunk 8]
```diff
@@ -32,26 +32,30 @@ import (
...
+	gctScriptManager         *gctscript.GctScriptManager
+	WebsocketRoutineManager  *WebsocketRoutineManager
+	WithdrawManager          *WithdrawManager
+	dataHistoryManager       *DataHistoryManager
+	currencyStateManager     *CurrencyStateManager
+	Settings                 Settings
+	uptime                   time.Time
+	GRPCShutdownSignal       chan struct{}
+	runtimeCtx               context.Context //nolint:containedctx // runtime-scoped cancellation context for startup + subsystem propagation
+	runtimeCancel            context.CancelFunc
```
Comment (shazbert) [inline engine/engine.go:55]: suggestion: Any chance we can use Start with updated parameter with context and just use that without all the runtime ctx methods? Then if we need to stop it the caller just calls cancel and then the Stop method?

File: engine/exchange_manager.go [hunk 9]
```diff
@@ -146,9 +147,16 @@ func (m *ExchangeManager) Shutdown(shutdownTimeout time.Duration) error {
...
-		name string
-		key  string
-		err  error
+		name     string
+		key      string
+		err      error
+		duration time.Duration
+	}
+
+	type pendingShutdown struct {
```
Comment (shazbert) [inline engine/exchange_manager.go:156]: Potentially, this type can be merged to the above type. And the loops merged into one having both startedAt and duration times fields, so that duration.

File: engine/helpers.go [hunk 10]
```diff
@@ -125,6 +126,15 @@ func (bot *Engine) SetSubsystem(subSystemName string, enable bool) error {
 		return errNilConfig
 	}
 
+	if enable && bot.isRuntimeShutdownRequested() {
```
Comment (shazbert) [inline engine/helpers.go:129]: I think its better to have a ctx param on these helper funcs.

File: engine/helpers.go [hunk 11]
```diff
@@ -622,6 +632,7 @@ func (bot *Engine) GetExchangeCryptocurrencyDepositAddress(ctx context.Context,
 
 // GetAllExchangeCryptocurrencyDepositAddresses obtains an exchanges deposit cryptocurrency list
 func (bot *Engine) GetAllExchangeCryptocurrencyDepositAddresses() map[string]ExchangeDepositAddresses {
+	runtimeCtx := bot.getRuntimeContext()
```
Comment (shazbert) [inline engine/helpers.go:635]: Like above the ctx should be populated through the paramater

File: exchanges/bitfinex/bitfinex_websocket.go [hunk 12]
```diff
@@ -143,20 +143,20 @@ func (e *Exchange) WsConnect() error {
...
-			e.Websocket.SetCanUseAuthenticatedEndpoints(false)
+		} else {
+			e.Websocket.Wg.Add(1)
+			go e.wsReadData(ctx, e.Websocket.AuthConn)
+			err = e.WsSendAuth(ctx)
+			if err != nil {
+				log.Errorf(log.ExchangeSys,
+					"%v - authentication failed: %v\n",
+					e.Name,
+					err)
```
Comment (shazbert) [inline exchanges/bitfinex/bitfinex_websocket.go]: WE can one line these logs.

File: main.go [hunk 13]
```diff
@@ -143,21 +144,38 @@ func main() {
...
 	engine.Bot.Settings.PrintLoadedSettings()
 
+	engine.Bot.EnsureRuntimeContext()
+	var shutdownRequested atomic.Bool
+	go waitForInterrupt(settings.Shutdown, engine.Bot, &shutdownRequested)
+
 	if err = engine.Bot.Start(); err != nil {
+		if shutdownRequested.Load() {
+			// Startup may have been interrupted before full subsystem initialisation.
+			// Avoid calling Stop() on a partially-initialized engine.
```
Comment (shazbert) [inline main.go]:
```suggestion
			// Avoid calling Stop() on a partially-initialised engine.
```

**Review by shazbert — CHANGES_REQUESTED** [review]
thanks!

File: backtester/engine/grpcserver.go (hunk 5 above)
Comment (thrasher-) [reply backtester/engine/grpcserver.go]: Jarvis: Addressed. Startup now initialises proxy setup before launching gRPC `Serve`, startup returns immediately on proxy bind/setup failures, and failure cleanup closes listeners so they are not left open. Added regression coverage in `TestStartRPCServerProxyFailureStopsListener`. Commit: https://github.com/thrasher-/gocryptotrader/commit/bafeb07b9c14def269c0751b44d1fb641e50b5a5

File: backtester/main.go (hunk 7 above)
Comment (thrasher-) [reply backtester/main.go:233]: Jarvis: Good call-out. I am intentionally leaving this unchanged in this patch: pushing raw context cancellation through the startup error channel can turn normal shutdown into a startup failure signal. Current flow handles shutdown via context path; I suggest addressing this in a focused lifecycle follow-up so cancellation semantics stay explicit.

File: engine/engine.go (hunk 8 above)
Comment (thrasher-) [reply engine/engine.go:55]: Jarvis: Agreed this is a reasonable direction. I am deferring this in the current patch because it broadens lifecycle API surface (`Start(ctx, ...)` + cancel/Stop flow) beyond the scoped startup-safety fix here. Happy to do it as a dedicated follow-up refactor.

File: engine/exchange_manager.go (hunk 9 above)
Comment (thrasher-) [reply engine/exchange_manager.go:156]: Jarvis: This is a good cleanup idea, but I am deferring it for now to keep this patch tightly scoped to correctness/stability changes. I can include type/loop consolidation in the same follow-up refactor set as the runtime-context API simplification.

File: engine/helpers.go (hunk 10 above)
Comment (thrasher-) [reply engine/helpers.go:129]: Jarvis: Agreed on explicit `ctx` parameter threading as a clearer long-term pattern. I have deferred this in this patch to avoid widening scope, and will fold it into the follow-up lifecycle/context refactor.

File: engine/helpers.go (hunk 11 above)
Comment (thrasher-) [reply engine/helpers.go:635]: Jarvis: Same as above — explicit `ctx` threading is cleaner, but I have deferred this in the current scoped fix set and will include it in the follow-up lifecycle/context refactor.

File: backtester/engine/grpcserver.go (hunk 6 above)
Comment (thrasher-) [reply backtester/engine/grpcserver.go]: Jarvis: Addressed. Removed trailing `\n` usage in the touched gRPC proxy log paths to keep logger output formatting consistent. Commit: https://github.com/thrasher-/gocryptotrader/commit/bafeb07b9c14def269c0751b44d1fb641e50b5a5

File: exchanges/bitfinex/bitfinex_websocket.go (hunk 12 above)
Comment (thrasher-) [reply exchanges/bitfinex/bitfinex_websocket.go]: Jarvis: Addressed. Simplified this logging call to a single-line `log.Errorf(...)` form. Commit: https://github.com/thrasher-/gocryptotrader/commit/bafeb07b9c14def269c0751b44d1fb641e50b5a5

File: main.go (hunk 13 above)
Comment (thrasher-) [reply main.go]: Jarvis: Addressed. Applied the inline wording simplification for that comment path. Commit: https://github.com/thrasher-/gocryptotrader/commit/bafeb07b9c14def269c0751b44d1fb641e50b5a5

File: engine/engine.go [hunk 14]
```diff
@@ -554,17 +573,118 @@ func (bot *Engine) Start() error {
 		}
 	}
 
+	startSuccessful = true
 	return nil
 }
 
+func (bot *Engine) getRuntimeContext() context.Context {
```
Comment (gloriousCode) [inline engine/engine.go:580]: While none of the below functions are very complex. I don't see why you cannot add dedicated unit tests for each of them. Direct tests are important (before you say there is indirect coverage)

Comment (thrasher-) [reply engine/engine.go:580]: jarvis: Addressed in 8ea33da9. Added dedicated unit tests for the runtime helper block in `engine/engine_test.go`, covering `getRuntimeContext`, `getRuntimeCancel`, `cancelRuntimeContext`, `clearRuntimeContext`, `setRuntimeShutdownRequested`, `isRuntimeShutdownRequested`, `EnsureRuntimeContext`, and `RequestShutdown`, including the shutdown reuse and edge-case branches in `EnsureRuntimeContext()`. Focused tests, focused `-race`, and `golangci-lint run ./engine/...` pass. Full `go test ./engine -count=1` is still blocked by the unrelated Bitfinex ticker failure already present on `master`.

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks Dirvus
